### ISSUE TITLE
make pathfinder more useful

### ISSUE THREAD

**Issue by obneq** [minetest/minetest#2614]
uselessness of pathfinder is a common complaint of mob makers. here are two suggestions to make it better:
- remove 'line-of-sight-optimisation' done at the end of pathfinding
    https://github.com/minetest/minetest/blob/master/src/pathfinder.cpp#L313

- add possibility to avoid given type(s) of nodes when finding path.

i have done both locally and it results in usabe paths found in a much more expected format.
My solution is a quick hack which only works for Dijkstra algo so far.

**Comment by paramat** [comment]
The PR is now merged 9af7f38e3f077ec42fcc3086c82d2676c53c84a6

**Comment by obneq** [comment]
@paramat yes, thanks for the link. However i will not close this issue yet, as there are still some glaring uselessness factors in/around mt pathfinding.

**Comment by 4aiman** [comment]
It is useful enough. 
The only thing which is wrong about it would be "failed to update cost map" error.

**Comment by sapier** [comment]
remove 'line-of-sight-optimisation' done at the end of pathfinding - done!
Well even if you meant this to be good it causes major sideeffects. E.g. mobf uses paths as waypoints, as they have been designed. Your change causes stop and go movement as of waypoints are defined as positions where some movement change has to happen.
Btw why do you need redundant information in those paths? it's just causing lot of data to be copied that could me more then easy interpolated. And interpolating is way less work then the other way round. Position based movement generators have to interpolate non integer positions anyway.

**Comment by obneq** [comment]
@sapier the LOS optimisation will happily throw away a path when you can see the endpoint. Like when you are looking at a point over a big hole in the ground pathfinder will give you just a straight line across the hole...
